lib/src/config.cc

```diff
@@ -304,6 +311,88 @@ namespace hocon {
         return long_list;
     }
 
+    string config::get_units(string s) {
+        int i = s.length() - 1;
+        while (i >= 0 && isalpha(s[i])) {
+            i--;
+        }
+        return s.substr(i + 1);
+    }
+
+    milliseconds config::get_duration(string const& path) const {
+        auto v = get_value(path);
+        if (auto d = dynamic_pointer_cast<const config_double>(v)) {
+            throw bad_value_exception(*v->origin(), path, "Durations without units must be integer values");
+        } else if (auto l = dynamic_pointer_cast<const config_long>(v)) {
+            return milliseconds(l->long_value());
+        } else if (auto i = dynamic_pointer_cast<const config_int>(v)) {
+            return milliseconds(i->long_value());
+        } else if (auto str = dynamic_pointer_cast<const config_string>(v)) {
+            return milliseconds(parse_duration(str->transform_to_string(), str->origin(), path).count());
+        } else {
+            throw bad_value_exception(*v->origin(), path, "Value at '" + path + "' was not an integer or string.");
+        }
+    }
+
+    double config::get_duration(string const& path, time_unit unit) const {
+        auto timespan = get_duration(path);
+        switch (unit) {
+            case time_unit::MILLISECONDS:
+                return timespan.count();
+            case time_unit::SECONDS:
+                return timespan.count() / 1000.0;
+            case time_unit::MINUTES:
+                return timespan.count() / 60000.0;
+            case time_unit::HOURS:
+                return timespan.count() / 3600000.0;
+            case time_unit::DAYS:
+                return timespan.count() / 86400000.0;
+            default:
+                throw config_exception("Not a valid time_unit");
+        }
+    }
+
+    milliseconds config::parse_duration(string input, shared_origin origin_for_exception, string path_for_exception) {
+        boost::algorithm::trim(input);
+        string original_unit_string = get_units(input);
+        string unit_string = original_unit_string;
+        string number_string = boost::algorithm::trim_copy(input.substr(0, input.length() - unit_string.length()));
+
+        if (number_string.empty()) {
+            throw bad_value_exception(*origin_for_exception, path_for_exception, "No number in duration value '" + input + "'");
+        }
+
+        if (unit_string.length() > 2 && unit_string.back() != 's') {
+            unit_string += "s";
+        }
+
+        double number = 0;
+        try {
+            number = boost::lexical_cast<double>(number_string);
```

Comment:
This can lose precision in a very minor case: x86 systems (usually) support extended precision floating point, meaning they use 80-bits to represent doubles - 1 bit for sign, 15 bits for the exponent, 1 bit for whether it's '0.xxx' or '1.xxx', and 63 bits for the fraction.

So this is safe if we only support 63-bits of precision (int64_t), but not 64-bits (uint64_t). The following example shows the difference.

```
#include <boost/lexical_cast.hpp>
#include <string>
#include <iostream>

int main(int argc, char** argv)
{
    std::string s = "18446744073709551615";
    double number = boost::lexical_cast<double>(s);
    uint64_t integer = static_cast<uint64_t>(number);
    std::cout << s << " != " << integer << std::endl;

    s = "9223372036854775807";
    number = boost::lexical_cast<double>(s);
    integer = static_cast<uint64_t>(number);
    std::cout << s << " == " << integer << std::endl;
    return 0;
}
```